app/controllers/application_controller.rb

```diff
@@ -1,8 +1,55 @@
 class ApplicationController < ActionController::Base
   include Pundit
-  
+
   protect_from_forgery with: :exception
   before_action :configure_permitted_parameters, if: :devise_controller?
+  #before_action :reset_session
+  before_action :create_guest
+  before_action :currert_shopping_cart
+
+
+
+
+
+  def currert_shopping_cart
+    
+    if user_signed_in?
+      if session[:shopping_cart]
+        @shopping_cart= session[:shopping_cart]
+        session[:shopping_cart]=@shopping_cart
+      else
+        @shopping_cart=Cart.create(user_id:current_user.id)
+      end
+    else
+
+      if session[:shopping_cart]
+        @shopping_cart= session[:shopping_cart]
+      else
+        @shopping_cart=Cart.new(user_id:session[:guest_user_id])
+        @shopping_cart.save(validate: false)
+        session[:shopping_cart]=@shopping_cart
+      end
+    end
+
+  end
+
+  def create_guest
+
+    if !(user_signed_in? || session[:guest])
+
+      session[:guest_user_id]=save_guest.id
```

Comment:
I think for every request. This will create a new user.Please review your approach.

Reply:
No way because **session[:guest]** will be false only when the user will login and when the user is logged in user_signed_in? will be true so the guest user will not be created only once when both are false( when user logout or new user visits).